Launch Services application information is sometimes not set

pvollan opened the pull request:
#### 33a1bca206e4f13118d2d3587297f47132fe48dd
<pre>
Launch Services application information is sometimes not set
<a href="https://bugs.webkit.org/show_bug.cgi?id=288861">https://bugs.webkit.org/show_bug.cgi?id=288861</a>
<a href="https://rdar.apple.com/145876451">rdar://145876451</a>

Reviewed by Chris Dumez.

There is currently a race between WebContent process checkin with Launch Services, and setting the
Launch Services application information for the WebContent process. Before the application information
can be set, the WebContent process needs to be checked in with Launch Services from the Networking
process. If Launch Services has not received the checkin message before it receives the application
information message, the latter will fail. This patch resolves this by storing the application
information when it&apos;s too soon to send, and then send it when the WebContent process has been
checked in.

* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/mac/NetworkConnectionToWebProcessMac.mm:
(WebKit::NetworkConnectionToWebProcess::updateActivePages):
(WebKit::NetworkConnectionToWebProcess::checkInWebProcess):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::ensureNetworkProcessConnection):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::updateProcessName):

Canonical link: <a href="https://commits.webkit.org/291461@main">https://commits.webkit.org/291461@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/042efa219fb30caacd81e201b0e691d8fc5d8eee

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/92852 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/12403 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/2131 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/97848 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/43376 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/12683 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/20855 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71023 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28451 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/95854 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/84055 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51351 "Passed tests") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/42691 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/1665 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/99870 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/19904 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/14601 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80042 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/20155 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/79954 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79342 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23880 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/1188 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/12930 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14858 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/19888 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/25064 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/19575 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/23035 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/21316 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
```diff
@@ -131,7 +131,7 @@ struct DataKey;
 
 class NetworkConnectionToWebProcess final
-    : public RefCounted<NetworkConnectionToWebProcess>
+    : public ThreadSafeRefCounted<NetworkConnectionToWebProcess>
```
Please use `ThreadSafeRefCounted<NetworkConnectionToWebProcess, WTF::DestructionThread::MainRunLoop>`. It is super dangerous otherwise. I don't think it'd be safe to destroy this on a non-main thread.

pvollan replied:
Ah, great catch, I have addressed these issues in the latest PR.

Thanks for reviewing!

cdumez commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
```diff
@@ -542,6 +542,9 @@ class NetworkConnectionToWebProcess final
 
     HashMap<WebTransportSessionIdentifier, Ref<NetworkTransportSession>> m_networkTransportSessions;
+#if ENABLE(LAUNCHSERVICES_SANDBOX_EXTENSION_BLOCKING)
+    std::optional<String> m_pendingDisplayName;
```
Case in point, it holds a String so it can only be destroyed on the thread where this string was allocated.

cdumez replied:
Also, we usually do not use std::optional with `String` since it is already nullable.

cdumez commented on Source/WebKit/NetworkProcess/mac/NetworkConnectionToWebProcessMac.mm:
```diff
@@ -105,11 +109,19 @@
     RELEASE_LOG(Process, "Launch Services checkin starting, infoDictionary = %{public}@", (__bridge NSDictionary *)infoDictionary.get());
 
-    Boolean ok = _LSApplicationCheckInProxyPtr()(kLSDefaultSessionID, auditToken.auditToken(), infoDictionary.get(), ^(CFDictionaryRef result, CFErrorRef error) {
+    auto block = makeBlockPtr([protectedThis = Ref { *this }, this, auditToken = auditToken](CFDictionaryRef result, CFErrorRef error) mutable {
```
` = auditToken` shouldn't be needed?

cdumez commented on Source/WebKit/NetworkProcess/mac/NetworkConnectionToWebProcessMac.mm:
```diff
@@ -105,11 +109,19 @@
         NSDictionary *dictionary = (__bridge NSDictionary *)result;
         RELEASE_LOG(Process, "Launch Services checkin completed, result = %{public}@, error = %{public}@", dictionary, (__bridge NSError *)error);
+        callOnMainRunLoop([protectedThis = WTFMove(protectedThis), this, auditToken] mutable {
+            if (m_pendingDisplayName) {
```
```
if (auto pendingDisplayName = std::exchange(m_pendingDisplayName, std::nullopt)) {
```

cdumez commented on Source/WebKit/NetworkProcess/mac/NetworkConnectionToWebProcessMac.mm:
```diff
@@ -105,11 +109,19 @@
         RELEASE_LOG(Process, "Launch Services checkin completed, result = %{public}@, error = %{public}@", dictionary, (__bridge NSError *)error);
+        callOnMainRunLoop([protectedThis = WTFMove(protectedThis), this, auditToken] mutable {
+            if (m_pendingDisplayName) {
+                updateActivePages(*m_pendingDisplayName, { }, WTFMove(auditToken));
```
`*pendingDisplayName `

cdumez commented on Source/WebKit/NetworkProcess/mac/NetworkConnectionToWebProcessMac.mm:
```diff
@@ -105,11 +109,19 @@
+        callOnMainRunLoop([protectedThis = WTFMove(protectedThis), this, auditToken] mutable {
+            if (m_pendingDisplayName) {
+                updateActivePages(*m_pendingDisplayName, { }, WTFMove(auditToken));
+                m_pendingDisplayName = std::nullopt;
```
No longer needed.

cdumez commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -1281,8 +1281,13 @@ NetworkProcessConnection& WebProcess::ensureNetworkProcessConnection()
-        if (auto auditToken = auditTokenForSelf())
+        if (auto auditToken = auditTokenForSelf()) {
             m_networkProcessConnection->protectedConnection()->send(Messages::NetworkConnectionToWebProcess::CheckInWebProcess(*auditToken), 0);
+            if (m_pendingDisplayName) {
```
Same as above about using `std::exchange()`

cdumez commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -668,19 +668,18 @@ static void setVideoDecoderBehaviors(OptionSet<VideoDecoderBehavior> videoDecode
         return;
     }
+#if ENABLE(LAUNCHSERVICES_SANDBOX_EXTENSION_BLOCKING)
+    m_pendingDisplayName = displayName;
```
Can we `WTFMove()`?

cdumez commented on Source/WebKit/WebProcess/WebProcess.h:
```diff
@@ -916,6 +916,9 @@ class WebProcess final : public AuxiliaryProcess {
     HashMap<String, int> m_notifyTokens;
 #endif
+#if ENABLE(LAUNCHSERVICES_SANDBOX_EXTENSION_BLOCKING)
+    std::optional<String> m_pendingDisplayName;
```
We normally never use `std::optional` with `String` since String has a null state.

cdumez commented on Source/WebKit/NetworkProcess/mac/NetworkConnectionToWebProcessMac.mm:
```diff
@@ -47,6 +47,10 @@
 #if USE(APPLE_INTERNAL_SDK)
     auto asn = adoptCF(_LSCopyLSASNForAuditToken(kLSDefaultSessionID, auditToken.auditToken()));
+    if (!asn) {
+        m_pendingDisplayName = overrideDisplayName;
```
This may be worth a code comment to explain why we're doing this.

cdumez commented on Source/WebKit/NetworkProcess/mac/NetworkConnectionToWebProcessMac.mm:
```diff
@@ -105,11 +113,19 @@
     RELEASE_LOG(Process, "Launch Services checkin starting, infoDictionary = %{public}@", (__bridge NSDictionary *)infoDictionary.get());
 
-    Boolean ok = _LSApplicationCheckInProxyPtr()(kLSDefaultSessionID, auditToken.auditToken(), infoDictionary.get(), ^(CFDictionaryRef result, CFErrorRef error) {
+    auto block = makeBlockPtr([protectedThis = Ref { *this }, this, auditToken](CFDictionaryRef result, CFErrorRef error) mutable {
```
Have you considered capturing a WeakPtr instead in the lambda? Then we wouldn't need ThreadSafeRefCounted at all.

I just don't want to encourage people to use this object off the main thread since it is really not meant to be.

pvollan replied:
That is a great point, fixed in latest PR.

Thanks for reviewing!

cdumez commented on Source/WebKit/NetworkProcess/mac/NetworkConnectionToWebProcessMac.mm:
```diff
@@ -105,11 +113,19 @@
+        callOnMainRunLoop([protectedThis = WTFMove(protectedThis), this, auditToken = WTFMove(auditToken)] mutable {
+            if (m_pendingDisplayName.isEmpty())
+                return;
+            updateActivePages(std::exchange(m_pendingDisplayName, emptyString()), { }, WTFMove(auditToken));
```
Should be using `String()`, not `emptyString()`

cdumez replied:
Also, I would have written this like so:
```
if (auto name = std::exchange(m_pendingDisplayName, String()))
    updateActivePages(name, { }, WTFMove(auditToken));
```

pvollan replied:
I tried this, but got a compile error saying that the conversion from String to bool was ambiguous, so I kept as is for now.

cdumez commented on Source/WebKit/NetworkProcess/mac/NetworkConnectionToWebProcessMac.mm:
```diff
@@ -105,11 +113,19 @@
         RELEASE_LOG(Process, "Launch Services checkin completed, result = %{public}@, error = %{public}@", dictionary, (__bridge NSError *)error);
+
+        callOnMainRunLoop([protectedThis = WTFMove(protectedThis), this, auditToken = WTFMove(auditToken)] mutable {
+            if (m_pendingDisplayName.isEmpty())
```
Should be using `isNull()`, not `isEmpty()`.

cdumez commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -1281,8 +1281,11 @@ NetworkProcessConnection& WebProcess::ensureNetworkProcessConnection()
-        if (auto auditToken = auditTokenForSelf())
+        if (auto auditToken = auditTokenForSelf()) {
             m_networkProcessConnection->protectedConnection()->send(Messages::NetworkConnectionToWebProcess::CheckInWebProcess(*auditToken), 0);
+            if (!m_pendingDisplayName.isEmpty())
```
`isNull()`

cdumez commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -1281,8 +1281,11 @@ NetworkProcessConnection& WebProcess::ensureNetworkProcessConnection()
+        if (auto auditToken = auditTokenForSelf()) {
             m_networkProcessConnection->protectedConnection()->send(Messages::NetworkConnectionToWebProcess::CheckInWebProcess(*auditToken), 0);
+            if (!m_pendingDisplayName.isEmpty())
+                m_networkProcessConnection->protectedConnection()->send(Messages::NetworkConnectionToWebProcess::UpdateActivePages(std::exchange(m_pendingDisplayName, emptyString()), { }, *auditToken), 0);
```
`emptyString()` -> `String()`

cdumez commented on Source/WebKit/NetworkProcess/mac/NetworkConnectionToWebProcessMac.mm:
```diff
@@ -105,11 +113,22 @@
     RELEASE_LOG(Process, "Launch Services checkin starting, infoDictionary = %{public}@", (__bridge NSDictionary *)infoDictionary.get());
 
-    Boolean ok = _LSApplicationCheckInProxyPtr()(kLSDefaultSessionID, auditToken.auditToken(), infoDictionary.get(), ^(CFDictionaryRef result, CFErrorRef error) {
+    auto block = makeBlockPtr([weakThis = WeakPtr { *this }, this, auditToken](CFDictionaryRef result, CFErrorRef error) mutable {
```
You're not allowed to capture `this` with `weakThis` per Safer CPP guidelines.

cdumez commented on Source/WebKit/NetworkProcess/mac/NetworkConnectionToWebProcessMac.mm:
```diff
@@ -105,11 +113,22 @@
         NSDictionary *dictionary = (__bridge NSDictionary *)result;
         RELEASE_LOG(Process, "Launch Services checkin completed, result = %{public}@, error = %{public}@", dictionary, (__bridge NSError *)error);
+
+        callOnMainRunLoop([weakThis = WTFMove(weakThis), this, auditToken = WTFMove(auditToken)] mutable {
```
ditto.

cdumez commented on Source/WebKit/NetworkProcess/mac/NetworkConnectionToWebProcessMac.mm:
```diff
@@ -47,6 +47,14 @@
+#if ENABLE(LAUNCHSERVICES_SANDBOX_EXTENSION_BLOCKING)
+        // In this case, the WebContent process has not been checked in with Launch Services yet.
+        // We store the display name, and set it after checkin has completed.
+        m_pendingDisplayName = overrideDisplayName;
```
The `overrideDisplayName` parameter for this function can be a `String&&` and you can use `WTFMove()` here.

cdumez reviewed (approved):
LGTM but it'd be good to do the `WTFMove()` I suggested before landing.

pvollan committed:
Launch Services application information is sometimes not set
https://bugs.webkit.org/show_bug.cgi?id=288861
rdar://145876451

Reviewed by Chris Dumez.

There is currently a race between WebContent process checkin with Launch Services, and setting the
Launch Services application information for the WebContent process. Before the application information
can be set, the WebContent process needs to be checked in with Launch Services from the Networking
process. If Launch Services has not received the checkin message before it receives the application
information message, the latter will fail. This patch resolves this by storing the application
information when it's too soon to send, and then send it when the WebContent process has been
checked in.

* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/mac/NetworkConnectionToWebProcessMac.mm:
(WebKit::NetworkConnectionToWebProcess::updateActivePages):
(WebKit::NetworkConnectionToWebProcess::checkInWebProcess):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::ensureNetworkProcessConnection):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::updateProcessName):

Canonical link: https://commits.webkit.org/291461@main
